### PR TITLE
Associate label sets with wall profiles

### PR DESCRIPTION
**Pull request by nsavoire**
Based on https://github.com/DataDog/pprof-nodejs/pull/98

Adds ability to associate custom label sets to wall samples. This can then be used by dd-trace-js to associate span ids with said samples.

Implementation notes
=================
The implementation adds JS API to set and clear labels. In practice, the `WallProfiler` JS class will gain getters and setters for properties `labels`.

Internally, the implementation uses a PPROF signal handler executing before V8's signal handler used for profiling. It will associate the current label set with a timestamp, and put it in a preallocated buffer (so allocations are avoided in the signal handler.) This happens once every 10ms while profiling. The buffer is sized generously, for 2x the number of samples for a single profile. 

When the profile is post-processed once every minute, the labels are matched to samples based on the timestamp, and emitted into the serialized profile.

Timestamp matching
=================

Two timestamps are collected in profiler signal handler and are associated with the current label: one before v8 handler is invoked and one after.
These two timestamps allow easily matching v8 profiling samples to labels, each sample is matched with the label whose timestamp interval contains the sample timestamp.
v8 does not expose publicly its clock API (`v8::base::TimeTicks::Now()`), matching samples / labels without a common clock source proved to be difficult, hence profiler uses private v8 API by redeclaring it.

### CODE REVIEW

File: bindings/profilers/wall.cc [hunk 1]
```diff
@@ -25,213 +28,609 @@
...
+    bool withLabels = info[3].As<Boolean>()->Value();
+
+#ifndef DD_WALL_USE_SIGPROF
+    if (withLabels) {
+      return Nan::ThrowTypeError("Labels are not supported.");
+    }
+#endif
+
+    if (includeLines && withLabels) {
+      // Currently custom labels are not compatible with caller line
```
Comment (r1viollet) [inline bindings/profilers/wall.cc:609]: great comment, should we add a ticket to v8 for this ?

File: bindings/profilers/wall.cc [hunk 2]
```diff
@@ -25,213 +28,609 @@
...
-           Nan::New<Number>(profile->GetStartTime()));
-  Nan::Set(js_profile,
-           Nan::New<String>("endTime").ToLocalChecked(),
-           Nan::New<Number>(profile->GetEndTime()));
-  return js_profile;
-}
+};
 
-WallProfiler::WallProfiler(int interval) : samplingInterval(interval) {}
+LabelSetsByNode WallProfiler::GetLabelSetsByNode(CpuProfile* profile,
```
Comment (r1viollet) [inline bindings/profilers/wall.cc:461]: Should we get a metric out of the function on the number of labeled samples ? How can we know if this works well when it runs in staging ?

File: bindings/buffer.hh [hunk 3]
```diff
@@ -0,0 +1,70 @@
+#pragma once
```
Comment (r1viollet) [inline bindings/buffer.hh:1]: minor: what should headers look like ?

**Review by r1viollet — DISMISSED** [review]
LGTM
A very well thought approach

File: bindings/profilers/wall.cc (hunk 2 above)
Comment (nsavoire) [reply bindings/profilers/wall.cc:461]: Only labelled samples ends up in the emitted profile, so we should indirectly see if it's working

File: bindings/buffer.hh (hunk 3 above)
Comment (nsavoire) [reply bindings/buffer.hh:1]: Don't know, should I use the same one as ddprof ?

File: bindings/profilers/wall.cc (hunk 2 above)
Comment (r1viollet) [reply bindings/profilers/wall.cc:247]: :+1: This makes it easy for long lived applications, though not easy for shorter applications (if we don't know how long the application ran for).

**Review by r1viollet — APPROVED** [review]
LGTM

File: bindings/profilers/wall.cc [hunk 4]
```diff
@@ -14,224 +14,409 @@
...
+  auto contextIt = contexts.begin();
+
+  // deltaIdx is the offset of the sample to process compared to current
+  // iteration index
+  int deltaIdx = 0;
+
+  // skip first sample because it's the one taken on profiler start, outside of
+  // signal handler
+  for (int i = 1; i < sampleCount; i++) {
+    // Handle out-of-order samples, hypothesis is that at most 2 consecutive
```
Comment (theanarkh) [inline bindings/profilers/wall.cc:268]: @nsavoire Hi, I want to know why this happened and why do you need to deal with this situation ? I read the V8 source code and found that the timestamp of samples is incremental (test/cctest/test-cpu-profiler.cc).

```c++
  uint64_t end_time = profile->GetEndTime();
  uint64_t current_time = profile->GetStartTime();
  CHECK_LE(current_time, end_time);
  for (int i = 0; i < profile->GetSamplesCount(); i++) {
    CHECK(profile->GetSample(i));
    uint64_t timestamp = profile->GetSampleTimestamp(i);
    // <=
    CHECK_LE(current_time, timestamp);
    CHECK_LE(timestamp, end_time);
    current_time = timestamp;
  }
```

**Pull request merged**